packages/next/src/server/future/route-matcher-managers/default-route-matcher-manager.ts

```diff
@@ -228,6 +229,11 @@ export class DefaultRouteMatcherManager implements RouteMatcherManager {
       return matcher.match(pathname, options)
     }
 
+    // For app page routes, we just ignore the locale if it's present.
+    if (matcher instanceof AppPageRouteMatcher) {
```

Comment:
@timneutkens So since the router has no idea what route a request will match, when `i18n` is set, it passes `(pathname = '/en-US/foo', options = { i18n: { pathname: '/foo', locale: '/en-US' } })` to all matchers to see if they can match.

And when that is passed to an AppPageRouteMatcher, it will miss. And we should instead pass `options.i18n?.pathname` (the original pathname) in that case.

The reason this only fail for dynamic app paths is, we have some matching logic in the router code for dynamic paths (I guess to avoid accidentally matching i18n as part of the route?) that we won't try to match the pathname without locales being set.

Reply:
I think there's a way to fix this in the `AppPageRouteMatcher` directly rather than patching it at this higher level. I can take a peek!

Reply:
This fix breaks the case where there is a route like:

```
app/[locale]/post/[slug]/page.js
```

Where locale management is done in app directory. The underlying issue is because the default locale is injected into paths higher up:

https://github.com/vercel/next.js/blob/9ad1f321b7902542acd2be041fb2f15f023a0ed9/packages/next/src/server/base-server.ts#L687-L689

So the fix would require this beheviour to be stripped out. Instead we should rely on a tweak to the locale detection. Today in the router, we assume the default locale correctly via the `inferDefaultLocale` option:

https://github.com/vercel/next.js/blob/9ad1f321b7902542acd2be041fb2f15f023a0ed9/packages/next/src/server/router.ts#L185-L195

This leaves the underlying `pathname` unchanged but detects the correct locale for the matchers. The incomming request pathname should always include the cleaned pathname (which still includes locale information).